pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package couchdbstore
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	_ "github.com/go-kivik/couchdb" // The CouchDB driver
+	"github.com/go-kivik/kivik"
+
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents an CouchDB implementation of the storage.Provider interface
+type Provider struct {
+	hostURL       string
+	couchDBClient *kivik.Client
+}
+
+const blankHostErrMsg = "hostURL for new CouchDB provider can't be blank"
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client}, nil
+}
+
+// OpenStore opens and returns a store for the given name.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	exists, err := p.couchDBClient.DBExists(context.Background(), name)
+	if err != nil {
+		return nil, err
+	}
+
+	if exists {
+		db := p.couchDBClient.DB(context.Background(), name)
+
+		if dbErr := db.Err(); dbErr != nil {
+			return nil, dbErr
+		}
+
+		return &CouchDBStore{db: db}, nil
+	}
+
+	err = p.couchDBClient.CreateDB(context.Background(), name)
+	if err != nil {
+		return nil, err
+	}
+
+	db := p.couchDBClient.DB(context.Background(), name)
+
+	if err := db.Err(); err != nil {
+		return nil, err
+	}
+
+	return &CouchDBStore{db: db}, nil
+}
+
+// CloseStore closes a previously opened store.
+// CouchDB and Kivik don't really have "close" mechanisms for individual databases, so there's nothing to do here.
+func (p *Provider) CloseStore(name string) error {
+	return nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.couchDBClient = nil
```

Comment:
[Client.Close()](https://godoc.org/github.com/go-kivik/kivik#Client.Close)?

Reply:
@llorllale The older version of Kivik I used initially didn't have that call, so that's why I didn't use it before. I see the newer one does - I've updated the method.